In [1]:
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import warnings
warnings.filterwarnings("ignore")

# Display all columns
pd.set_option('display.max_columns', None)

In [2]:
# Get working directory
import os
print('Current directory is: ', os.getcwd())

Current directory is:  c:\Users\shubh\Desktop\Dissertation\project\Backend


In [3]:
# Load the geo-encoded property data
df = pd.read_csv(r'Output Files\\properties_with_coordinates.csv') 
df.head()

ID   Possession Status Availability Starts From Floor No Commercial  \
0  12684       Ready to Move                      NaN       20          Y   
1  12683       Ready to Move                      NaN       18          N   
2  12682  Under Construction                  Dec '25        5          N   
3  12681  Under Construction                  Dec '24        8          Y   
4  12680  Under Construction                  Mar '23       16          N   

                               Developer Approved Authority Name  \
0  TATA Housing Development Company Ltd.                     TMC   
1                  Sai Satyam Developers                    KDMC   
2                          Birla Estates                    KDMC   
3                      Godrej Properties                     NaN   
4                          Tycoons Group                  ABAAut   

   Units Available      Price Price (English) Flooring Type  \
0             10.0  6300000.0          63 Lac     Vitrified   
1              0.0  5400000.0          54 Lac     Vitrified   
2             70.0  9000000.0          90 Lac     Vitrified   
3              NaN  4950000.0        49.5 Lac           NaN   
4              2.0  6570000.0        65.7 Lac     Vitrified   

  Electricity Status      Maintenance Type  Maintenance Charges  \
0   No/Rare Powercut  Per sq. Unit Monthly                  3.0   
1   No/Rare Powercut               Monthly               1200.0   
2   No/Rare Powercut               Monthly               3200.0   
3                NaN                   NaN                  NaN   
4   No/Rare Powercut               Monthly                  2.0   

   Booking Amount                                           Landmark  \
0        100000.0                                        Rajoli naka   
1        100000.0    This property has reputed scholl ints vicinity.   
2        100000.0  Shahad is one of the attractive locations to o...   
3        500000.0                             majiwada metro station   
4         50000.0  Close to tree house school, Dmart, KDMC garden...   

   Covered Area                Project Name  sqft Price   Carpet Area  \
0         579.0                Tata Amantra      10880.0        579.0   
1         850.0            Sai Satyam Homes       6352.0        585.0   
2        1050.0                 Birla Vanya       8571.0        815.0   
3         561.0              Godrej Nirvaan       8824.0        419.0   
4        1067.0  Tycoons Solitaire Sapphire       6157.0        667.0   

     Area Name   Property Uniqueness Unit of Carpet Area Society  \
0  Kalyan West                   NaN               Sq-ft       Y   
1  Kalyan West  East Facing Property               Sq-ft       Y   
2  Kalyan West                   NaN               Sq-ft       Y   
3  Kalyan West                   NaN               Sq-ft       Y   
4  Kalyan West                   NaN               Sq-ft       Y   

  Ownership Type  furnished Type  Bathroom            Parking        Facing  \
0       Freehold     Unfurnished       2.0          1 Covered          East   
1       Freehold     Unfurnished       2.0             1 Open          East   
2       Freehold     Unfurnished       2.0  1 Covered, 1 Open          East   
3       Freehold  Semi-Furnished       2.0                NaN          East   
4       Freehold     Unfurnished       2.0          1 Covered  North - East   

               Amenities Facing Transaction Type      Floor Data  \
0                     Main Road     New Property  20 of 34 Floor   
1        Garden/Park, Main Road     New Property  18 of 18 Floor   
2  Pool, Garden/Park, Main Road     New Property   5 of 27 Floor   
3                   Garden/Park     New Property   8 of 26 Floor   
4        Garden/Park, Main Road     New Property  16 of 18 Floor   

  Type of Property   City  bedroom Data Referred From  floors  \
0        Apartment  Thane        2              Agent    34.0   
1        Apartment  Thane        2              Agent    

In [4]:
# Check shape
df.shape

(8154, 150)

In [5]:
# Print column names
for column in df.columns:
    print(column)

ID
Possession Status
Availability Starts From
Floor No
Commercial
Developer
Approved Authority Name
Units Available
Price
Price (English)
Flooring Type
Electricity Status
Maintenance Type
Maintenance Charges
Booking Amount
Landmark
Covered Area
Project Name
sqft Price 
Carpet Area
Area Name
Property Uniqueness
Unit of Carpet Area
Society
Ownership Type
furnished Type
Bathroom
Parking
Facing
Amenities Facing
Transaction Type
Floor Data
Type of Property
City
bedroom
Data Referred From
floors
Water Status
Luxury Flat
Covered Area.1
Carpet Area Unit
isPrimeLocationProperty
Tenants Preference
Location
Property
covArea Unit
isLuxuryServiceProvided
Property Lifespan
balconies
Rera
Power Back Up
Lift
Rain Water Harvesting
Club House
Swimming Pool
Gymnasium
Park
Parking.1
Security
Water Storage
Private Terrace/Garden
Vaastu Compliant
Service/Goods Lift
Air Conditioned
Visitor Parking
Intercom Facility
Maintenance Staff
Waste Disposal
Laundry Service
Internet/Wi-Fi Connectivity
DTH Television Fa

In [6]:
# Format column names
df.columns = df.columns.str.strip()

In [7]:
# Drop the not required columns
df = df.drop(columns=[
    'ID',
    'Price (English)',
    'Ownership Type',
    'Possession By',
    'isLuxuryServiceProvided',
    'Approved Authority Name',
    'Rera',
    'Location',
    'Units Available',
    'Property',
    'Property Uniqueness',
    'Tenants Preference',
    'Amenities Facing',
    'Covered Area.1',
    'Booking Amount',
    'Land Area Unit',
    'Land Area / Covered Area',
    'geocoding_success',
    'address_variant_used',
    'Land Area / Covered Area',
    'Maintenance Type',
    'Maintenance Charges',
    'Pantry Type',
    'Society',
    'Electricity Status',
    'Luxury Flat',
    'Water Status',
    'NRI Pref',
    'Parking',
    'Carpet Area Unit',
    'Floor Data',
    'Flooring Type',
    'Pantry Type'
])

In [8]:
# Check updated shape
df.shape

(8154, 119)

# <b>Data Assesment</b>

In [9]:
pd.set_option('display.max_columns', None)
df.head(2)

Possession Status Availability Starts From Floor No Commercial  \
0     Ready to Move                      NaN       20          Y   
1     Ready to Move                      NaN       18          N   

                               Developer      Price  \
0  TATA Housing Development Company Ltd.  6300000.0   
1                  Sai Satyam Developers  5400000.0   

                                          Landmark  Covered Area  \
0                                      Rajoli naka         579.0   
1  This property has reputed scholl ints vicinity.         850.0   

       Project Name  sqft Price  Carpet Area    Area Name Unit of Carpet Area  \
0      Tata Amantra     10880.0        579.0  Kalyan West               Sq-ft   
1  Sai Satyam Homes      6352.0        585.0  Kalyan West               Sq-ft   

  furnished Type  Bathroom Facing Transaction Type Type of Property   City  \
0    Unfurnished       2.0   East     New Property        Apartment  Thane   
1    Unfurnished       2.0   East     New Property        Apartment  Thane   

   bedroom Data Referred From  floors isPrimeLocationProperty covArea Unit  \
0        2              Agent    34.0                       Y        Sq-ft   
1        2              Agent    18.0                       Y        Sq-ft   

  Property Lifespan  balconies  Power Back Up  Lift  Rain Water Harvesting  \
0  New Construction        NaN              1     1                      1   
1  New Construction        NaN              1     1                      1   

   Club House  Swimming Pool  Gymnasium  Park  Parking.1  Security  \
0           1              1          1     1          1         1   
1           1              1          1     1          1         1   

   Water Storage  Private Terrace/Garden  Vaastu Compliant  \
0              1                       1                 1   
1              1                       1                 1   

   Service/Goods Lift  Air Conditioned  Visitor Parking  Intercom Facility  \
0                   1                1                1                  1   
1                   1                1                1                  1   

   Maintenance Staff  Waste Disposal  Laundry Service  \
0                  1               1                1   
1                  1               1                1   

   Internet/Wi-Fi Connectivity  DTH Television Facility  RO Water System  \
0                            1                        1                1   
1                            1                        1                1   

   Banquet Hall  Bar/Lounge  Cafeteria/Food Court  Conference Room  Piped Gas  \
0             1           1                     1                1          1   
1             1           1                     1                1          1   

   Jogging and Strolling Track  Outdoor Tennis Courts  Mansion  Downtown  \
0                            1                      1        1         1   
1                            1                      1        1         1   

   Skyline View  Wrap Around Balcony  Island Kitchen Layout  Full Glass Wall  \
0             1                    1                      1                1   
1             1                    1                      1                1   

   House help accommodation  Concierge Services  Garden View  Marble flooring  \
0                         1                   1            1                1   
1                         1                   1            1                1   

   Well Furnished  Modular Kitchen  Helipad  Private pool  Private Jaccuzi  \
0               1                1        1             1                1   
1               1                1        1             1                1   

   Mini Cinema Theatre  Golf Course  Infinity Swimming Pool  \
0                    1            1                       1   
1                    1            1                       1   

   Pool with temperature control  Sea facing  Skydeck  Earth

In [10]:
# Check duplicate entries
df.drop(columns=['latitude', 'longitude', 'geocoded_address']).duplicated().sum()

np.int64(1832)

In [11]:
# Unique entries IDs
idx = df.drop(columns=['latitude', 'longitude', 'geocoded_address']).drop_duplicates(keep='first').index

# Construct unique properties dataframe
df = df.iloc[idx]

# Reset index
df.reset_index(inplace=True, drop=True)

In [12]:
# Check shape
df.shape

(6322, 119)

In [ ]:
# Info
pd.set_option('display.max_info_columns', 200)
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6322 entries, 0 to 6321
Data columns (total 119 columns):
 #    Column                                   Non-Null Count  Dtype  
---   ------                                   --------------  -----  
 0    Possession Status                        6276 non-null   object 
 1    Availability Starts From                 1874 non-null   object 
 2    Floor No                                 5921 non-null   object 
 3    Commercial                               6322 non-null   object 
 4    Developer                                5420 non-null   object 
 5    Price                                    6271 non-null   float64
 6    Landmark                                 4101 non-null   object 
 7    Covered Area                             5665 non-null   float64
 8    Project Name                             6322 non-null   object 
 9    sqft Price                               5704 non-null   float64
 10   Carpet Area                       

-------------

#### Column assesment: Possession Status

In [14]:
# Format values
df['Possession Status'] = df['Possession Status'].str.strip()

# Count unique values
df['Possession Status'].value_counts()

Possession Status
Ready to Move         4512
Under Construction    1612
Dec '26                 28
Dec '24                 21
Dec '25                 20
Dec '23                 17
Jun '24                 13
Jun '26                 12
Dec '28                  8
Dec '27                  8
Mar '25                  4
Sep '26                  3
Mar '24                  3
Immediately              3
Nov '25                  2
Sep '23                  2
Apr '28                  2
Jun '25                  2
Apr '24                  2
Nov '23                  1
Select Date              1
Name: count, dtype: int64

In [15]:
# Format
df['Possession Status'] = df['Possession Status'].str.replace(" '", " ")

In [16]:
# Replace past dates with 'Ready to move' status
df['Possession Status'].replace({'Immediately': 'Ready to Move', 
                                 'Dec 24': 'Ready to Move', 
                                 'Dec 23': 'Ready to Move', 
                                 'Jun 24': 'Ready to Move', 
                                 'Mar 25': 'Ready to Move',
                                 'Mar 24': 'Ready to Move',
                                 'Sep 23': 'Ready to Move',
                                 'Jun 25': 'Ready to Move',
                                 'Apr 24': 'Ready to Move',
                                 'Nov 23': 'Ready to Move'
                                 }, inplace=True)

In [17]:
# Updated value counts
df['Possession Status'].value_counts()

Possession Status
Ready to Move         4580
Under Construction    1612
Dec 26                  28
Dec 25                  20
Jun 26                  12
Dec 27                   8
Dec 28                   8
Sep 26                   3
Nov 25                   2
Apr 28                   2
Select Date              1
Name: count, dtype: int64

In [18]:
# Drop row having 'Select Date' Possesion status
index_to_drop = df[df['Possession Status'] == 'Select Date'].index
df.drop(index=index_to_drop, inplace=True)

# Check updated shape
df.shape


(6321, 119)

In [19]:
# Check missing values
df['Possession Status'].isna().sum()

np.int64(46)

----------

#### Column assesment: Availability Starts From

In [20]:
# Format values
df['Availability Starts From'] = df['Availability Starts From'].str.strip()

In [21]:
# 'Ready to Move' properties indexes
ready_to_move_indexes = df[df['Possession Status'] == 'Ready to Move'].index

# Set 'Availability Starts From' to 'Immediately' where Possesion status is 'Ready to Move'
df.loc[ready_to_move_indexes, 'Availability Starts From'] = 'Immediately'

In [22]:
# Count unique values
df['Availability Starts From'].value_counts()

Availability Starts From
Immediately    4580
Dec '25         300
Dec '24         282
Dec '23         219
Dec '26         141
Jun '23          66
Mar '23          59
Jun '24          55
Jun '25          54
Jun '26          39
Mar '25          39
Mar '24          33
Apr '23          26
Sep '23          22
Dec '27          18
Sep '25          17
May '25          16
Jul '25          15
Dec '28          15
Mar '26          14
Oct '23          14
Aug '23          13
Jul '24          10
Nov '23          10
Oct '25           9
Nov '24           9
May '23           9
May '24           9
Jan '24           6
Nov '25           6
Jun '27           6
Jul '23           5
Apr '24           5
Jan '26           5
Oct '26           5
Oct '24           5
Aug '25           4
Sep '24           4
Jan '25           4
Aug '24           4
Sep '26           4
Feb '25           3
Mar '28           3
Apr '25           3
Nov '26           2
Feb '24           2
Jan '27           2
Apr '26           2
Jun '28        

In [23]:
# Format values
df['Availability Starts From'] = df['Availability Starts From'].str.replace(" '", " ")

# Update past dates to 'Immediately'
df['Availability Starts From'].replace({ 
                                 'Dec 22': 'Immediately',

                                 'Mar 23': 'Immediately',
                                 'Apr 23': 'Immediately',
                                 'May 23': 'Immediately',
                                 'Jun 23': 'Immediately',
                                 'Jul 23': 'Immediately',
                                 'Aug 23': 'Immediately',
                                 'Sep 23': 'Immediately',
                                 'Oct 23': 'Immediately',
                                 'Nov 23': 'Immediately',
                                 'Dec 23': 'Immediately',

                                 'Jan 24': 'Immediately',
                                 'Feb 24': 'Immediately',
                                 'Mar 24': 'Immediately',
                                 'Apr 24': 'Immediately',
                                 'May 24': 'Immediately',
                                 'Jun 24': 'Immediately',
                                 'Jul 24': 'Immediately',
                                 'Aug 24': 'Immediately',
                                 'Sep 24': 'Immediately',
                                 'Oct 24': 'Immediately',
                                 'Nov 24': 'Immediately',
                                 'Dec 24': 'Immediately',

                                 'Jan 25': 'Immediately',
                                 'Feb 25': 'Immediately',
                                 'Mar 25': 'Immediately',
                                 'Apr 25': 'Immediately',
                                 'May 25': 'Immediately',
                                 'Jun 25': 'Immediately',
                                 'Jul 25': 'Immediately'
                                 }, inplace=True)

In [24]:
# Update 'Possesion Status' of properties to 'Ready to Move' where 'Availability Starts From' is 'Immediately'
immediate_availability_indexes = df[df['Availability Starts From'] == 'Immediately'].index
df.loc[immediate_availability_indexes, 'Possession Status'] = 'Ready to Move'

In [25]:
# Get properties where date is provided for Possesion status
possesion_status_date_provided_indexes = df[~((df['Possession Status'] == 'Ready to Move') | (df['Possession Status'] == 'Under Construction') | (df['Possession Status'].isna()))].index

# Copy these dates to 'Availability Starts From'
df.loc[possesion_status_date_provided_indexes, 'Availability Starts From'] = df.loc[possesion_status_date_provided_indexes, 'Possession Status']

In [26]:
# Replace future dates in 'Possesion Status' with 'Under Construction'
idx = df[(df['Possession Status'] != 'Ready to Move') & (df['Possession Status'] != 'Under Construction') & (df['Possession Status'].notna())].index
df.loc[idx, 'Possession Status'] = 'Under Construction'

In [27]:
# Replace NaN values in 'Possession Status' and 'Availability Starts From' with 'Not Available, Contact Developer'
df['Possession Status'].fillna('Not Available, Contact Developer', inplace=True)
df['Availability Starts From'].fillna('Not Available, Contact Developer', inplace=True)

In [28]:
# Check missing values in 'Possesion status' and 'Availability starts from' features
print('Missing values in Possesion status: ', df['Possession Status'].isna().sum())
print('Missing values in Availability Starts From: ', df['Availability Starts From'].isna().sum())

Missing values in Possesion status:  0
Missing values in Availability Starts From:  0


In [29]:
# Check shape
df.shape

(6321, 119)

--------------

#### Column assesment: Floor No

In [30]:
# Format
df['Floor No'] = df['Floor No'].str.strip()

In [31]:
# Count frequency of each unique value
freq_df = df['Floor No'].value_counts().reset_index()
freq_df.columns = ['Value', 'Frequency']

fig = px.bar(freq_df, x='Value', y='Frequency', title='Frequency Plot of Floor No')
fig.show()

In [32]:
# Convert Ground, Upper Basement, Lower Basement into numbers
df['Floor No'].replace({'Ground': '0', 'Upper Basement': '-1', 'Lower Basement': '-2'}, inplace=True)


In [33]:
# Convert to Numeric
df['Floor No'] = pd.to_numeric(df['Floor No'], errors='coerce')

In [34]:
# Check missing values
df['Floor No'].isna().sum()

np.int64(401)

In [35]:
# Check shape
df.shape

(6321, 119)

----------------

#### Column assesment: Commercial

In [36]:
# Format
df['Commercial'] = df['Commercial'].str.capitalize().str.strip()

In [37]:
# Count unique values
df['Commercial'].value_counts()

Commercial
N    5243
Y    1078
Name: count, dtype: int64

In [38]:
# Check missing values
df['Commercial'].isna().sum()

np.int64(0)

In [39]:
# Check shape
df.shape

(6321, 119)

-------------

#### Column assesment: Developer

In [40]:
# Format
df['Developer'] = df['Developer'].str.title().str.strip()

In [41]:
# Count unique values
df['Developer'].value_counts()

Developer
Lodha                         439
Rustomjee                     140
Kalpataru Ltd.                113
Bombay Realty                 112
Runwal Developers             105
                             ... 
Pnk Group                       1
Ritu Developer                  1
Sanghvi Group Of Companies      1
R V Constructions               1
Kakad Realty                    1
Name: count, Length: 937, dtype: int64

In [42]:
# Check missing values
df['Developer'].isna().sum()

np.int64(901)

In [43]:
# Check shape
df.shape

(6321, 119)

-----------

#### Column assesment: Price

In [44]:
# Convert prices into Crores
df['Price'] = df['Price']/10000000

# Rename column
df.rename(columns={'Price': 'Price (Crores)'}, inplace=True)

In [45]:
# Check missing values
df['Price (Crores)'].isna().sum()

np.int64(50)

In [46]:
# Check shape
df.shape

(6321, 119)

---------------

#### Column assesment: Landmark

In [47]:
df['Landmark'] = df['Landmark'].str.capitalize().str.strip()

In [48]:
# Check missing values
df['Landmark'].isna().sum()

np.int64(2220)

In [49]:
# Check shape
df.shape

(6321, 119)

--------------------

#### Column assesment: Covered Area, covArea Unit

In [50]:
# Check missing values in 'Covered Area'
df['Covered Area'].isna().sum()

np.int64(657)

In [51]:
# Rename 'covArea Unit' to 'Unit of Covered Area'
df.rename(columns={'covArea Unit': 'Unit of Covered Area'}, inplace=True)

# Missing values in 'Unit of Covered Area'
df['Unit of Covered Area'].isna().sum()

np.int64(656)

In [52]:
# Count unique values in 'Unit of Covered Area'
df['Unit of Covered Area'].value_counts()

Unit of Covered Area
Sq-ft    5657
Sq-m        6
Bigha       1
Acre        1
Name: count, dtype: int64

In [53]:
# Show properties having Covered area units in Sq-m or Bigha or Acre
df[(df['Unit of Covered Area'] == 'Sq-m') | (df['Unit of Covered Area'] == 'Bigha') | (df['Unit of Covered Area'] == 'Acre')]

Possession Status          Availability Starts From  Floor No  \
360        Ready to Move                       Immediately       1.0   
1132       Ready to Move                       Immediately       4.0   
1199       Ready to Move                       Immediately      12.0   
1666       Ready to Move                       Immediately      15.0   
3156       Ready to Move                       Immediately       2.0   
5142  Under Construction  Not Available, Contact Developer      17.0   
5614       Ready to Move                       Immediately       1.0   
6038       Ready to Move                       Immediately       9.0   

     Commercial                     Developer  Price (Crores)  \
360           Y  Satellite Builders Pvt. Ltd.           1.350   
1132          Y                           NaN           0.320   
1199          N            Vinay Unique Group           0.340   
1666          N             Godrej Properties             NaN   
3156          N                           NaN           1.250   
5142          Y                     Uk Realty           0.678   
5614          N                           NaN           0.300   
6038          N             Deepak Developers           1.400   

                         Landmark  Covered Area            Project Name  \
360                           NaN         560.0       Satellite Classic   
1132                          NaN         630.0     Virar Bolinj Shakti   
1199                   Near dmart          36.0     Vinay Unique Corner   
1666                          NaN          50.0          Godrej Emerald   
3156                          NaN          64.0  The Ascent Apartment B   
5142                          NaN          35.0              UK Iridium   
5614  Ycce college within 150mtr.          50.0        Suyash Apartment   
6038                          NaN          35.0        Deepak Daffodils   

      sqft Price  Carpet Area        Area Name Unit of Carpet Area  \
360          1.0        415.0  Jogeshwari East               Bigha   
1132         0.0        600.0       Virar West                Acre   
1199      8774.0          NaN       Virar West                Sq-m   
1666         NaN         45.0    Bhayandarpada                Sq-m   
3156     18145.0        484.0   Santacruz East                Sq-m   
5142     17996.0          NaN   Kandivali East                Sq-m   
5614     60000.0        750.0     Andheri West                Sq-m   
6038     37160.0          NaN     Andheri East                Sq-m   

      furnished Type  Bathroom Facing Transaction Type Type of Property  \
360      Unfurnished       2.0  South           Resale        Apartment   
1132     Unfurnished       2.0    NaN           Resale        Apartment   
1199  Semi-Furnished       2.0   East           Resale        Apartment   
1666     Unfurnished       2.0    NaN           Resale        Apartment   
3156     Unfurnished       2.0  North           Resale        Apartment   
5142     Unfurnished       2.0    NaN     New Property        Apartment   
5614     Unfurnished       2.0   East           Resale        Apartment   
6038     Unfurnished       1.0    NaN           Resale        Apartment   

        City  bedroom Data Referred From  floors isPrimeLocationProperty  \
360   Mumbai        1              Owner     7.0                       N   
1132  Mumbai        2              Owner     4.0                       N   
1199  Mumbai        1              Owner    14.0                       N   
1666   Thane        2              Owner    25.0                       N   
3156  Mumbai        1              Owner     7.0                       N   
5142  Mumbai        1            Builder    33.0                       Y   
5614  Nagpur        2              Owner     4.0                       Y   
6038  Mumbai        1              Owner    11.0                       N   

     Unit of Covered Area  Property Lifespan  balconies  Power Back Up  Lift  \
360                 Bigh

In [54]:
# Data Correction
df.loc[360, 'Unit of Covered Area'] = 'Sq-ft'
df.loc[1132, 'Unit of Covered Area'] = 'Sq-ft'

df.loc[[1199, 1666, 3156, 5142, 5614, 6038], 'Covered Area'] = df.loc[[1199, 1666, 3156, 5142, 5614, 6038], 'Covered Area'] * 10.7639
df.loc[[1199, 1666, 3156, 5142, 5614, 6038], 'Unit of Covered Area'] = 'Sq-ft'

In [55]:
# Check shape
df.shape

(6321, 119)

----------------

#### Column assesment: Project Name

In [56]:
# Rename
df.rename(columns={'Project Name': 'Society'}, inplace=True)

In [57]:
# Format
df['Society'] = df['Society'].str.title().str.strip()

In [58]:
# Count unique values
df['Society'].value_counts()

Society
Lodha Amara                  77
Rustomjee Crown              56
Kalpataru Vivant             52
Bombay Dyeing Two Icc        51
Island City Center           36
                             ..
Mahindra Happinest Kalyan     1
Shirin Apartment              1
Green Fields                  1
Mahakali Chs                  1
Crystal Enclave               1
Name: count, Length: 2334, dtype: int64

In [59]:
# Check missing values
df['Society'].isna().sum()

np.int64(0)

In [60]:
# Drop 2 societies named Sion East (Invalid Society name caught manually)
idx = df[df['Society'] == 'Sion East'].index

# Drop permanently
df.drop(index=idx, inplace=True)

In [61]:
# Check shape
df.shape

(6319, 119)

---------------

#### Column assesment: sqft Price

In [62]:
# Rename column
df.rename(columns={'sqft Price': 'sqft Price (INR)'}, inplace=True)

In [63]:
# Inaccurate entries where 'sqft Price' is not matching with Price/Covered_area value
idx = df[abs(((df['Price (Crores)'] / df['Covered Area']) * 10000000).round() - df['sqft Price (INR)']) > 1][['Price (Crores)', 'Covered Area', 'sqft Price (INR)']].index

# Data Correction
df.loc[idx, 'sqft Price (INR)'] = ((df.loc[idx, 'Price (Crores)'] * 10000000) / df.loc[idx, 'Covered Area']).round()

print('Number of listings corrected: ', len(idx))

Number of listings corrected:  208


In [64]:
# Check missing values
df['sqft Price (INR)'].isna().sum()

np.int64(617)

In [65]:
# Check shape
df.shape

(6319, 119)

---------------

#### Column assesment: Carpet Area, Unit of Carpet Area

In [66]:
# Check missing values
df['Carpet Area'].isna().sum()

np.int64(1130)

In [67]:
# Inaccurate entries: Properties having Carpet area greater than the Covered area
df_sub = df[df['Carpet Area'] > df['Covered Area']]
print('Properties having Carpet area greater than the Covered area: ', len(df_sub))

Properties having Carpet area greater than the Covered area:  43


In [68]:
# Draw 5 random samples
df_sub[['Covered Area', 'Unit of Covered Area', 'Carpet Area', 'Unit of Carpet Area', 'bedroom']].sample(10)

Covered Area Unit of Covered Area  Carpet Area Unit of Carpet Area  \
4859         990.0                Sq-ft       1300.0               Sq-ft   
1775         475.0                Sq-ft       1500.0               Sq-ft   
4489         560.0                Sq-ft        650.0               Sq-ft   
4131         300.0                Sq-ft        350.0               Sq-ft   
5681         480.0                Sq-ft        580.0               Sq-ft   
3526         540.0                Sq-ft        555.0               Sq-ft   
3892         700.0                Sq-ft        920.0               Sq-ft   
2194         910.0                Sq-ft       1200.0               Sq-ft   
4705        1540.0                Sq-ft       2500.0               Sq-ft   
6110         870.0                Sq-ft        950.0               Sq-ft   

      bedroom  
4859        3  
1775        1  
4489        1  
4131        1  
5681        2  
3526        1  
3892        2  
2194        3  
4705        3  
6110        2

##### Observation: 
This appears to be a data entry error. The Covered area and Carpet area seems to have been accidently swapped for these 43 properties.

In [69]:
# Data Correction: Swap the values
original_covered_area = df_sub['Covered Area']
df_sub['Covered Area'] = df_sub['Carpet Area']
df_sub['Carpet Area'] = original_covered_area

In [70]:
# Update the original dataframe
df.update(df_sub)

In [71]:
# Count unique values in 'Unit of Carpet Area'
df['Unit of Carpet Area'].value_counts()

Unit of Carpet Area
Sq-ft    5655
Sq-m        6
Bigha       1
Acre        1
Name: count, dtype: int64

In [72]:
# Properties having Carpet area units in Sq-m or Bigha or Acre
df_sub = df[(df['Unit of Carpet Area'] == 'Sq-m') | (df['Unit of Carpet Area'] == 'Bigha') | (df['Unit of Carpet Area'] == 'Acre')]
df_sub

Possession Status          Availability Starts From  Floor No  \
360        Ready to Move                       Immediately       1.0   
1132       Ready to Move                       Immediately       4.0   
1199       Ready to Move                       Immediately      12.0   
1666       Ready to Move                       Immediately      15.0   
3156       Ready to Move                       Immediately       2.0   
5142  Under Construction  Not Available, Contact Developer      17.0   
5614       Ready to Move                       Immediately       1.0   
6038       Ready to Move                       Immediately       9.0   

     Commercial                     Developer  Price (Crores)  \
360           Y  Satellite Builders Pvt. Ltd.           1.350   
1132          Y                           NaN           0.320   
1199          N            Vinay Unique Group           0.340   
1666          N             Godrej Properties             NaN   
3156          N                           NaN           1.250   
5142          Y                     Uk Realty           0.678   
5614          N                           NaN           0.300   
6038          N             Deepak Developers           1.400   

                         Landmark  Covered Area                 Society  \
360                           NaN      560.0000       Satellite Classic   
1132                          NaN      630.0000     Virar Bolinj Shakti   
1199                   Near dmart      387.5004     Vinay Unique Corner   
1666                          NaN      538.1950          Godrej Emerald   
3156                          NaN      688.8896  The Ascent Apartment B   
5142                          NaN      376.7365              Uk Iridium   
5614  Ycce college within 150mtr.      750.0000        Suyash Apartment   
6038                          NaN      376.7365        Deepak Daffodils   

      sqft Price (INR)  Carpet Area        Area Name Unit of Carpet Area  \
360            24107.0      415.000  Jogeshwari East               Bigha   
1132            5079.0      600.000       Virar West                Acre   
1199            8774.0          NaN       Virar West                Sq-m   
1666               NaN       45.000    Bhayandarpada                Sq-m   
3156           18145.0      484.000   Santacruz East                Sq-m   
5142           17996.0          NaN   Kandivali East                Sq-m   
5614            5574.0      538.195     Andheri West                Sq-m   
6038           37160.0          NaN     Andheri East                Sq-m   

      furnished Type  Bathroom Facing Transaction Type Type of Property  \
360      Unfurnished       2.0  South           Resale        Apartment   
1132     Unfurnished       2.0    NaN           Resale        Apartment   
1199  Semi-Furnished       2.0   East           Resale        Apartment   
1666     Unfurnished       2.0    NaN           Resale        Apartment   
3156     Unfurnished       2.0  North           Resale        Apartment   
5142     Unfurnished       2.0    NaN     New Property        Apartment   
5614     Unfurnished       2.0   East           Resale        Apartment   
6038     Unfurnished       1.0    NaN           Resale        Apartment   

        City  bedroom Data Referred From  floors isPrimeLocationProperty  \
360   Mumbai        1              Owner     7.0                       N   
1132  Mumbai        2              Owner     4.0                       N   
1199  Mumbai        1              Owner    14.0                       N   
1666   Thane        2              Owner    25.0                       N   
3156  Mumbai        1              Owner     7.0                       N   
5142  Mumbai        1            Builder    33.0                       Y   
5614  Nagpur        2              Owner     4.0                       Y   
6038  Mumbai        1              Owner    11.0                       N   

     Unit of Covered Area  Property Lifespan  balc

In [73]:
# Update 'Unit of Carpet area' to 'Sq-ft'
df_sub['Unit of Carpet Area'] = 'Sq-ft'

# Update the original dataframe
df.update(df_sub)

In [74]:
# Check missing values in 'Unit of Carpet Area'
df['Unit of Carpet Area'].isna().sum()

np.int64(656)

In [75]:
# Check shape
df.shape

(6319, 119)

--------------

#### Column assesment: Area Name

In [76]:
# Rename column
df.rename(columns={'Area Name': 'Area'}, inplace=True)

In [77]:
# Format
df['Area'] = df['Area'].str.title().str.strip()

In [78]:
# Count unique values
df['Area'].value_counts()

Area
Dombivli East       388
Andheri West        340
Andheri East        283
Goregaon East       241
Kandivali West      233
                   ... 
Azad Nagar Jjc        1
Parsi Wada            1
Mogra Village         1
Maheshwari Nagar      1
Marol Maroshi         1
Name: count, Length: 272, dtype: int64

In [79]:
# Study City-Society-Area together
df_society_area = df.groupby(['City', 'Society', 'Area']).size().reset_index(name='Number of Listings')
df_society_area.head(10)

City            Society                              Area  \
0   Bhiwandi  Sanjivani Complex                        Thane West   
1  Hyderabad   Orchid Residency            Venkatapuram-Balapuram   
2     Mumbai  10 Central Avenue                    Santacruz West   
3     Mumbai      111 Hyde Park                        Malad East   
4     Mumbai       127 Raj Home                      Mahajan Wadi   
5     Mumbai       127 Raj Home                    Mira Road East   
6     Mumbai       127 Raj Home                        Penkarpada   
7     Mumbai           19 North                    Kandivali West   
8     Mumbai           19 North  Mahatama Gandhi Road Amrut Nagar   
9     Mumbai           25 South                    Century Bazaar   

   Number of Listings  
0                   1  
1                   1  
2                   1  
3                   1  
4                   3  
5                   3  
6                  11  
7                   1  
8                   1  
9                   1

For multiple societies same area is depicted at different granular level in same city e.g. Property listings in 127 Raj Home society Mumbai appears in multiple rows with different granular areas of same area. 

In [80]:
# Construct dataframe for societies which are depicting different areas at different granular level
df_societies_with_diff_granular_areas = df_society_area[df_society_area[['City', 'Society']].duplicated(keep=False)]
df_societies_with_diff_granular_areas

City                 Society                              Area  \
4     Mumbai            127 Raj Home                      Mahajan Wadi   
5     Mumbai            127 Raj Home                    Mira Road East   
6     Mumbai            127 Raj Home                        Penkarpada   
7     Mumbai                19 North                    Kandivali West   
8     Mumbai                19 North  Mahatama Gandhi Road Amrut Nagar   
...      ...                     ...                               ...   
2824   Thane          Vihang Vermont                        Thane West   
2835   Thane       Vrindavan Complex                          Majiwada   
2836   Thane       Vrindavan Complex                 Vrindavan Society   
2841   Thane  Wadhwa Mahek Residency                       Kalyan East   
2842   Thane  Wadhwa Mahek Residency                    Nandivali Gaon   

      Number of Listings  
4                      3  
5                      3  
6                     11  
7                      1  
8                      1  
...                  ...  
2824                   1  
2835                   2  
2836                   1  
2841                   1  
2842                   1  

[917 rows x 4 columns]

In [81]:
# Export to excel for analysis and correction 
df_societies_with_diff_granular_areas.to_excel('Output Files//societies_with_diff_granular_areas.xlsx', index=False)

In [82]:
# Load manually processed societies_with_diff_granular_areas.xlsx file
df_societies_areas = pd.read_excel("Output Files//manually_processed_societies_with_diff_granular_areas.xlsx")
df_societies_areas.sample(5)

Society    Area_Broader    City
96    Godrej Tranquil  Kandivali East  Mumbai
23   Aquila At Sarova  Kandivali East  Mumbai
147   Lalani Grandeur   Goregaon East  Mumbai
358     Garden Estate      Thane West   Thane
381     Mali Pinnacle     Kalyan East   Thane

In [83]:
# Drop City-Society duplicate entries
idx = df_societies_areas[df_societies_areas[['Society', 'City']].duplicated()].index
df_societies_areas.drop(index=idx, inplace=True)

In [84]:
# Merge df with df_societies_areas to get Broader Area of societies
df = df.merge(df_societies_areas, how='left', on=['Society', 'City'])
df.head()

Possession Status Availability Starts From  Floor No Commercial  \
0       Ready to Move              Immediately      20.0          Y   
1       Ready to Move              Immediately      18.0          N   
2  Under Construction                   Dec 25       5.0          N   
3       Ready to Move              Immediately       8.0          Y   
4       Ready to Move              Immediately      16.0          N   

                               Developer  Price (Crores)  \
0  Tata Housing Development Company Ltd.           0.630   
1                  Sai Satyam Developers           0.540   
2                          Birla Estates           0.900   
3                      Godrej Properties           0.495   
4                          Tycoons Group           0.657   

                                            Landmark  Covered Area  \
0                                        Rajoli naka         579.0   
1    This property has reputed scholl ints vicinity.         850.0   
2  Shahad is one of the attractive locations to o...        1050.0   
3                             Majiwada metro station         561.0   
4  Close to tree house school, dmart, kdmc garden...        1067.0   

                      Society  sqft Price (INR)  Carpet Area         Area  \
0                Tata Amantra           10880.0        579.0  Kalyan West   
1            Sai Satyam Homes            6352.0        585.0  Kalyan West   
2                 Birla Vanya            8571.0        815.0  Kalyan West   
3              Godrej Nirvaan            8824.0        419.0  Kalyan West   
4  Tycoons Solitaire Sapphire            6157.0        667.0  Kalyan West   

  Unit of Carpet Area  furnished Type  Bathroom        Facing  \
0               Sq-ft     Unfurnished       2.0          East   
1               Sq-ft     Unfurnished       2.0          East   
2               Sq-ft     Unfurnished       2.0          East   
3               Sq-ft  Semi-Furnished       2.0          East   
4               Sq-ft     Unfurnished       2.0  North - East   

  Transaction Type Type of Property   City  bedroom Data Referred From  \
0     New Property        Apartment  Thane        2              Agent   
1     New Property        Apartment  Thane        2              Agent   
2     New Property        Apartment  Thane        2              Agent   
3     New Property        Apartment  Thane        1              Agent   
4     New Property        Apartment  Thane        2              Agent   

   floors isPrimeLocationProperty Unit of Covered Area Property Lifespan  \
0    34.0                       Y                Sq-ft  New Construction   
1    18.0                       Y                Sq-ft  New Construction   
2    27.0                       Y                Sq-ft               NaN   
3    26.0                       Y                Sq-ft               NaN   
4    18.0                       Y                Sq-ft               NaN   

   balconies  Power Back Up  Lift  Rain Water Harvesting  Club House  \
0        NaN              1     1                      1           1   
1        NaN              1     1                      1           1   
2        3.0              1     1                      1           1   
3        NaN              1     1                      1           1   
4        1.0              1     1                      1           1   

   Swimming Pool  Gymnasium  Park  Parking.1  Security  Water Storage  \
0              1          1     1          1         1              1   
1              1          1     1          1         1              1   
2              1          1     1          1         1              1   
3              1          1     1          1         1              1   
4              1          1     1          1         1              1   

   Private Terrace/Garden  Vaastu Compliant  Service/Goods Lift  \
0                       1                 1                   1   
1                       1        

In [85]:
# Get IDs where Area Broader is not null
idx = df[df['Area_Broader'].notna()].index

# Update Area to Area Broader for these listings
df.loc[idx, 'Area'] = df.loc[idx, 'Area_Broader']

In [86]:
# Update Area Broader to Area for manual updation of Area Broader
df['Area_Broader'] = df['Area']

In [87]:
# Value counts in Area Broader
df['Area_Broader'].value_counts()

Area_Broader
Andheri West         446
Dombivli East        420
Thane West           390
Andheri East         370
Kandivali East       370
                    ... 
Dn Nagar               1
Gondavali Gaothan      1
Jb Nagar               1
Jijamata Road          1
Hanuman Nagar          1
Name: count, Length: 197, dtype: int64

In [88]:
# Update Area Broader manually
df['Area_Broader'] = df['Area_Broader'].replace({
    'Majiwada': 'Thane West',
    'Balkum Village': 'Thane West',
    'Manpada': 'Thane West',
    'Kolshet Road': 'Thane West',
    'Pokharan Road Number 2': 'Thane West',
    'Hiranandani Estate': 'Thane West',
    'Bhayandarpada': 'Thane West',
    'Waghbil': 'Thane West',
    'Dhokali': 'Thane West',
    'Anand Nagar Thane West': 'Thane West',
    'Vartak Nagar': 'Thane West',
    'Patlipada': 'Thane West',
    'Kandarpada': 'Dahisar West',
    'Kolshet': 'Thane West',
    'Teen Hath Naka': 'Thane West',
    'Kopri': 'Thane East',
    'Owale': 'Thane West',
    'Thakur Village, Kandivali East': 'Kandivali East',
    'Marol': 'Andheri East',
    'Mahavir Nagar': 'Kandivali West',
    'Panch Pakhdi': 'Thane West',
    'Ashok Van': 'Dahisar East',
    'Vasant Vihar': 'Thane West',
    'Thakur Complex': 'Kandivali East',
    'Kapurbawdi': 'Thane West',
    'Kalina': 'Santacruz East',
    'Seven Bungalows': 'Andheri West',
    'Charkop': 'Kandivali West',
    'Asha Nagar': 'Kandivali East',
    'Amboli': 'Andheri West',
    'Sakinaka Junction': 'Andheri East',
    '4 Bunglows': 'Andheri West',
    'Naupada': 'Thane West',
    'Lokhandwala Complex': 'Andheri West',
    'Mahajan Wadi': 'Kandivali East',
    'Charai': 'Thane West',
    'Shivaji Park': 'Dadar West',
    'Veera Desai Road': 'Andheri West',
    'Natwar Nagar': 'Jogeshwari East',
    'Indira Nagar': 'Andheri East',
    'Rawal Pada': 'Dahisar East',
    'Wagle Industrial Estate': 'Thane West',
    'Subhash Nagar': 'Jogeshwari East',
    'Film City Road': 'Goregaon East',
    'Irani Wada': 'Dahisar East',
    'D.N. Nagar': 'Andheri West',
    'Vaishali Nagar': 'Dahisar East',
    'Chakki Naka': 'Kalyan West',
    'Brhmanwada': 'Matunga',
    'Dindoshi': 'Goregaon East',
    'Saki Vihar Road': 'Andheri East',
    'Chakala': 'Andheri East',
    'Lokhandwala Twp': 'Kandivali East',
    'Chikhal Dongari': 'Virar West',
    'Sher E Punjab Society': 'Andheri East',
    'Laxmi Nagar': 'Goregaon West',
    'Gokuldham': 'Goregaon East',
    'Pokharan Road Number 1': 'Thane West',
    'Miragaon': 'Miragaon, Mira Road',
    'Marol Naka': 'Andheri East',
    'Khar Danda': 'Khar West',
    'Marol Maroshi Road': 'Andheri East',
    'Poonam Nagar': 'Andheri East',
    'Versova Marg': 'Andheri West',
    'Alika Nagar': 'Mira Road East',
    'Gilbert Hill Jjc': 'Andheri West',
    'Charkop Sector 6': 'Kandivali West',
    'Azad Nagar': 'Andheri West',
    'Louis Wadi': 'Thane West',
    'Lokmanya Nagar': 'Thane West',
    'Mhatre Wadi': 'Kandivali West',
    'Nandivli': 'Dombivli East',
    'Chulne': 'Vasai West',
    'Y K Nagar': 'Virar West',
    'Bolinj ': 'Virar West',
    'Dhanukar Wadi': 'Kandivali West',
    'Samel Pada': 'Nalasopara East',
    'Vijay Nagar': 'Kalyan East',
    'Sen Nagar': 'Nalasopara East',
    'Hindu Colony': 'Dadar East',
    'Pandurang Wadi': 'Goregaon East',
    'Shree Siddhivinayak Mandir Chowk': 'Prabhadevi',
    'Penkarpada': 'Mira Road East',
    'Shivai Nagar': 'Thane West',
    'Jakhadevi Chowk': 'Vasai West',
    'Irla': 'Vile Parle West',
    'Tirupati Nagar': 'Virar West',
    'Kaka Sahab Dadarkar Chowk': 'Dadar West',
    'Golani Naka': 'Vasai East',
    'Dhumal Nagar': 'Vasai East',
    'Hind Nagar': 'Nalasopara East',
    'Amrut Nagar': 'Ghatkopar West',
    'Desai Village': 'Thane East',
    'Anand Nagar': 'Dahisar East',
    'Leo Peter Wadi': 'Vasai West',
    'Maratha Colony Dahisar': 'Dahisar East',
    'Balaji Nagar': 'Mira Road East',
    'Gokul Township': 'Virar West',
    'Agashi': 'Virar West',
    'Virat Nagar': 'Virar West',
    'Sonar Pada': 'Dombivli East',
    'Nandivali Gaon': 'Dombivli East',
    'Khopat': 'Thane West',
    'Meenatai Thakrey Chowk': 'Thane West',
    'Brahmand': 'Thane West',
    'Lok Puram': 'Thane West',
    'Kolbad': 'Thane West',
    'Ram Maruti Road Naupada': 'Thane West',
    'Sundar Nagar, Kalina': 'Santacruz East',
    'Laxmiben Chheda Nagar': 'Ghatkopar West',
    'Bimbisar Nagar': 'Goregaon East',
    'Century Bazaar': 'Prabhadevi',
    'Runwal Nagar': 'Thane West',
    'North Ave': 'Santacruz West',
    'L Tilak Colony': 'Dadar East',
    'Dadar Tt Circle': 'Dadar East',
    'Mahalaxmi Sindhi Colony': 'Mahalaxmi',
    'Jayprakash Nagar': 'Goregaon East',
    'Aarey Milk Colony': 'Goregaon East',
    'Ganesh Nagar': 'Borivali East',
    'Royal Palms Estate': 'Goregaon East',
    'Kanya Pada': 'Goregaon East',
    'Pathanwadi': 'Malad East',
    'Shivaji Nagar Malad East': 'Malad East',
    'Marine Drive': 'Churchgate',
    'Five Gardens': 'Matunga East',
    'Kopargaon': 'Dombivli West',
    'Upper Govind Nagar': 'Malad East',
    'Rpf Colony': 'Dombivli East',
    'Asavari': 'Mahim',
    'Mahatama Gandhi Road Amrut Nagar': 'Ghatkopar West',
    'Charkop Sector 8': 'Kandivali West',
    'Charkop Sector 9': 'Kandivali West',
    'Charkop Sector 3': 'Kandivali West',
    'Golibar': 'Santacruz East',
    'Old Khar': 'Khar West',
    'Samata Nagar': 'Kandivali East',
    'Akurli Nagar': 'Kandivali East',
    'Four Bungalows': 'Andheri West',
    'Millat Nagar': 'Andheri West',
    'Lallu Bhai Park': 'Andheri West',
    'Dn Nagar': 'Andheri West',
    'Gondavali Gaothan': 'Andheri West',
    'Jb Nagar': 'Andheri East',
    'Jijamata Road': 'Andheri East',
    'Hanuman Nagar': 'Worli'
})

In [89]:
# Updated value counts in Area Broader
df['Area_Broader'].value_counts()

Area_Broader
Thane West                       1044
Andheri West                      490
Dombivli East                     425
Andheri East                      410
Kandivali East                    407
Goregaon East                     336
Kandivali West                    319
Malad East                        211
Virar West                        188
Dahisar East                      182
Dadar West                        169
Dadar East                        151
Naigaon East                      135
Kalyan West                       123
Mira Road East                    118
Santacruz West                    110
Santacruz East                    108
Prabhadevi                        103
Vasai East                        101
Khar West                          84
Kalyan East                        75
Jogeshwari West                    70
Sion Koliwada                      62
Matunga West                       60
Chembur East                       59
Oshiwara                           58

In [90]:
# Areas in Thane city
df[df['City'] == 'Thane']['Area_Broader'].value_counts().sort_index()

Area_Broader
Dombivli East     424
Dombivli West      21
Kalyan East        74
Kalyan West       123
Thane East         18
Thane West       1042
Name: count, dtype: int64

In [91]:
# Areas in Mumbai city
df[df['City'] == 'Mumbai']['Area_Broader'].value_counts().sort_index()

Area_Broader
Agripada                          12
Andheri East                     409
Andheri West                     487
Bolinj                             2
Borivali East                      1
Chembur East                      59
Chembur West                      31
Churchgate                         2
Dadar East                       151
Dadar West                       169
Dahisar East                     182
Dahisar West                      41
Dombivli East                      1
Ghatkopar West                     3
Goregaon East                    336
Goregaon West                      3
Grant Road                         7
Jogeshwari East                   58
Jogeshwari Vikhroli Link Road     57
Jogeshwari West                   70
Kalyan East                        1
Kandivali East                   407
Kandivali West                   318
Khar West                         84
Lower Parel East                   3
Lower Parel West                  42
Mahalaxmi                

In [92]:
# Mumbai city listings having Thane West or Dombivli East or Kalyan East areas [Inaccurate Broader areas]
df[(df['City'] == 'Mumbai') & (
    (df['Area_Broader'] == 'Thane West') |
    (df['Area_Broader'] == 'Dombivli East') |
    (df['Area_Broader'] == 'Kalyan East') 

)]


Possession Status Availability Starts From  Floor No Commercial  \
1741     Ready to Move              Immediately       6.0          N   
4725     Ready to Move              Immediately       8.0          N   
6164     Ready to Move              Immediately       3.0          N   

               Developer  Price (Crores)  \
1741  Gaurang Associates            2.25   
4725               Lodha            1.00   
6164        Hubtown Ltd.           12.00   

                                               Landmark  Covered Area  \
1741                                                NaN        1400.0   
4725  The property is in palava phase - ii, dombivli...        1223.0   
6164               Just above big boy toys car showrrom        3350.0   

                     Society  sqft Price (INR)  Carpet Area         Area  \
1741               West View           16071.0          NaN      Naupada   
4725  Lodha Lakeshore Greens            8177.0        673.0   Rpf Colony   
6164         Hubtown Sunmist           35820.0       2600.0  Vijay Nagar   

     Unit of Carpet Area  furnished Type  Bathroom        Facing  \
1741               Sq-ft       Furnished       2.0           NaN   
4725               Sq-ft  Semi-Furnished       2.0          West   
6164               Sq-ft     Unfurnished       4.0  North - East   

     Transaction Type   Type of Property    City  bedroom Data Referred From  \
1741           Resale          Apartment  Mumbai        3              Owner   
4725           Resale          Apartment  Mumbai        2              Owner   
6164     New Property  Residential House  Mumbai        4              Owner   

      floors isPrimeLocationProperty Unit of Covered Area  Property Lifespan  \
1741     8.0                       N                Sq-ft                NaN   
4725    16.0                       Y                Sq-ft      5 to 10 years   
6164    15.0                       Y                Sq-ft  Less than 5 years   

      balconies  Power Back Up  Lift  Rain Water Harvesting  Club House  \
1741        NaN              0     0                      0           0   
4725        2.0              1     1                      1           1   
6164        4.0              1     1                      1           1   

      Swimming Pool  Gymnasium  Park  Parking.1  Security  Water Storage  \
1741              0          0     0          0         0              0   
4725              1          1     1          1         1              1   
6164              1          1     1          1         1              1   

      Private Terrace/Garden  Vaastu Compliant  Service/Goods Lift  \
1741                       0                 0                   0   
4725                       1                 1                   1   
6164                       1                 1                   1   

      Air Conditioned  Visitor Parking  Intercom Facility  Maintenance Staff  \
1741                0                0                  0                  0   
4725                1                1                  1                  1   
6164                1                1                  1                  1   

      Waste Disposal  Laundry Service  Internet/Wi-Fi Connectivity  \
1741               0                0                            0   
4725               1                1                            1   
6164               1                1                            1   

      DTH Television Facility  RO Water System  Banquet Hall  Bar/Lounge  \
1741                        0                0             0           0   
4725                        1                1             1           1   
6164                        1                1             1           1   

      Cafeteria/Food Court  Conference Room  Piped Gas  \
1741                     0                0          0   
4725                     1                1          1   
6164                     1                1          1   


In [93]:
# Manual correction
df.loc[1741, 'Area_Broader'] = 'Bandra East'
df.loc[4725, 'City'] = 'Thane'
df.loc[6164, 'Area_Broader'] = 'Andheri East'


In [94]:
# Check missing values
print('Missing values in Area: ', df['Area'].isna().sum())

Missing values in Area:  0


In [95]:
# Check shape
df.shape

(6319, 120)

-------------

#### Column assesment: furnished Type

In [96]:
# Rename column
df.rename(columns={'furnished Type': 'Furnished Type'}, inplace=True)

In [97]:
# Format
df['Furnished Type'] = df['Furnished Type'].str.title().str.strip()

In [98]:
# Count unique values
df['Furnished Type'].value_counts()

Furnished Type
Unfurnished       3431
Semi-Furnished    1993
Furnished          844
Name: count, dtype: int64

In [99]:
# Check missing values
df['Furnished Type'].isna().sum()

np.int64(51)

In [100]:
# Check shape
df.shape

(6319, 120)

---------------

#### Column assesment: Bathroom

In [101]:
# Rename column
df.rename(columns={'Bathroom': 'Bathrooms'}, inplace=True)

In [102]:
# Count unique values
df['Bathrooms'].value_counts()

Bathrooms
2.0     3560
3.0     1122
1.0     1025
4.0      384
5.0      127
6.0       50
7.0        7
10.0       1
Name: count, dtype: int64

In [103]:
# Check missing values
df['Bathrooms'].isna().sum()

np.int64(43)

In [104]:
# Check shape
df.shape

(6319, 120)

-----------------

#### Column assesment: Facing

In [105]:
# Format
df['Facing'] = df['Facing'].str.title().str.strip()

In [106]:
# Count unique values
df['Facing'].value_counts()

Facing
East            2716
North - East     665
West             501
North            394
North - West     119
South -West       91
South             88
South - East      65
Name: count, dtype: int64

In [107]:
# Format values
df['Facing'] = df['Facing'].str.replace(' - ', '-')
df['Facing'] = df['Facing'].str.replace(' -', '-')

df['Facing'].value_counts()

Facing
East          2716
North-East     665
West           501
North          394
North-West     119
South-West      91
South           88
South-East      65
Name: count, dtype: int64

In [108]:
# Missing values
df['Facing'].isna().sum()

np.int64(1680)

In [109]:
# Check shape
df.shape

(6319, 120)

----------------

#### Column assesment: Transaction Type

In [110]:
# Format
df['Transaction Type'] = df['Transaction Type'].str.title().str.strip()

In [111]:
# Count unique values
df['Transaction Type'].value_counts()

Transaction Type
Resale          3862
New Property    2450
Rent               3
Other              3
Rent/Lease         1
Name: count, dtype: int64

In [112]:
# Drop properties with 'Other' or 'Rent/Lease' transaction type
idx = df[(df['Transaction Type'] == 'Rent') | (df['Transaction Type'] == 'Other') | (df['Transaction Type'] == 'Rent/Lease')].index
df.drop(index = idx, inplace=True)

In [113]:
# Check missing values
df['Transaction Type'].isna().sum()

np.int64(0)

In [114]:
# Check shape
df.shape

(6312, 120)

---------------

#### Column assesment: Type of Property

In [115]:
# Format
df['Type of Property'] = df['Type of Property'].str.title().str.strip()

In [116]:
# Count unique values
df['Type of Property'].value_counts()

Type of Property
Apartment                  6201
Residential House            51
Builder Floor Apartment      29
Villa                        21
Penthouse                    10
Name: count, dtype: int64

In [117]:
# Check missing values
df['Type of Property'].isna().sum()

np.int64(0)

In [118]:
# Check shape
df.shape

(6312, 120)

---------------

#### Column assesment: City

In [119]:
# Format
df['City'] = df['City'].str.title().str.strip()

In [120]:
# Count unique values
df['City'].value_counts()

City
Mumbai       4604
Thane        1701
Nagpur          5
Bhiwandi        1
Hyderabad       1
Name: count, dtype: int64

In [121]:
# Drop properties where City is Nagpur, Bhiwandi or Hyderabad
idx = df[(df['City'] == 'Nagpur') | (df['City'] == 'Bhiwandi') | (df['City'] == 'Hyderabad')].index
df.drop(index=idx, inplace=True)

In [122]:
df.shape

(6305, 120)

----------------

#### Column assesment: bedroom

In [123]:
# Rename the column
df.rename(columns={'bedroom': 'Bedrooms'}, inplace=True)

In [124]:
# Count unique values
df['Bedrooms'].value_counts()

Bedrooms
2    2517
1    2019
3    1315
4     370
5      66
6      18
Name: count, dtype: int64

In [125]:
# Check missing values
df['Bedrooms'].isna().sum()

np.int64(0)

In [126]:
# Check shape
df.shape

(6305, 120)

---------------

#### Column assesment: Data Referred from

In [127]:
# Rename
df.rename(columns={'Data Referred From': 'Posted By'}, inplace=True)

In [128]:
# Format
df['Posted By'] = df['Posted By'].str.title().str.strip()

In [129]:
# Count unique values
df['Posted By'].value_counts()

Posted By
Agent      3387
Owner      2753
Builder     165
Name: count, dtype: int64

In [130]:
# Check missing values
df['Posted By'].isna().sum()


np.int64(0)

In [131]:
# Check shape
df.shape

(6305, 120)

------------

#### Column assesment: floors

In [132]:
# Rename the column
df.rename(columns={'floors': 'Total Floors'}, inplace=True)

In [133]:
# Count unique values
df['Total Floors'].value_counts()

Total Floors
7.0      694
22.0     403
23.0     272
8.0      249
15.0     226
        ... 
200.0      1
93.0       1
71.0       1
91.0       1
74.0       1
Name: count, Length: 82, dtype: int64

In [134]:
# Check missing values
df['Total Floors'].isna().sum()

np.int64(405)

In [135]:
# Check shape
df.shape

(6305, 120)

---------------

#### Column assesment: isPrimeLocationProperty

In [136]:
# Format
df['isPrimeLocationProperty'] = df['isPrimeLocationProperty'].str.capitalize().str.strip()

In [137]:
# Count unique values
df['isPrimeLocationProperty'].value_counts()

isPrimeLocationProperty
Y    4885
N    1420
Name: count, dtype: int64

In [138]:
# Check missing values
df['isPrimeLocationProperty'].isna().sum()

np.int64(0)

In [139]:
# Check shape
df.shape

(6305, 120)

-----------

#### Column assesment: Property Lifespan

In [140]:
# Format
df['Property Lifespan'] = df['Property Lifespan'].str.capitalize().str.strip()

In [141]:
# Count unique values
df['Property Lifespan'].value_counts()

Property Lifespan
New construction     1064
Less than 5 years     823
5 to 10 years         771
10 to 15 years        467
Above 20 years        249
15 to 20 years        238
Name: count, dtype: int64

In [142]:
# Check missing values
df['Property Lifespan'].isna().sum()

np.int64(2693)

In [143]:
# Check shape
df.shape

(6305, 120)

--------

#### Column assesment: balconies

In [144]:
# Rename
df.rename(columns={'balconies': 'Balconies'}, inplace=True)

In [145]:
# Count unique values
df['Balconies'].value_counts()

Balconies
1.0    1333
2.0     742
3.0     634
4.0     187
5.0      45
6.0      15
8.0       2
7.0       1
Name: count, dtype: int64

In [146]:
# Check missing values
df['Balconies'].isna().sum()

np.int64(3346)

In [147]:
# Check shape
df.shape

(6305, 120)

-------------

#### Column assesment: Parking.1

In [148]:
# Rename the column
df.rename(columns={'Parking.1': 'Parking'}, inplace=True)

In [149]:
# Count unique values
df['Parking'].value_counts()

Parking
1    4523
0    1782
Name: count, dtype: int64

In [150]:
# Check missing values
df['Parking'].isna().sum()

np.int64(0)

In [151]:
# Check shape
df.shape

(6305, 120)

--------------

In [153]:
# Save the dataframe in CSV format
df.to_csv('Output Files\\stage_2_property_data.csv', index=False)